File: android/src/main/kotlin/com/adyen/checkout/flutter/apiOnly/AdyenCSE.kt
package com.adyen.checkout.flutter.apiOnly

import com.adyen.checkout.cse.CardEncrypter
import com.adyen.checkout.flutter.generated.EncryptedCardDTO
import com.adyen.checkout.flutter.generated.UnencryptedCardDTO
import com.adyen.checkout.flutter.utils.ConfigurationMapper.toNativeUnencryptedCard

internal object AdyenCSE {
    fun encryptCard(
        card: UnencryptedCardDTO,
        publicKey: String,
    ): EncryptedCardDTO {
        val encryptedCard = CardEncrypter.encryptFields(card.toNativeUnencryptedCard(), publicKey)
        return EncryptedCardDTO(
            encryptedCardNumber = encryptedCard.encryptedCardNumber,
            encryptedExpiryMonth = encryptedCard.encryptedExpiryMonth,
            encryptedExpiryYear = encryptedCard.encryptedExpiryYear,
            encryptedSecurityCode = encryptedCard.encryptedSecurityCode,
        )
    }

    fun encryptBin(bin: String, publicKey: String): String =
        CardEncrypter.encryptBin(bin, publicKey)
}

File: ios/adyen_checkout/Sources/adyen_checkout/apiOnly/AdyenCSE.swift
import Adyen
import AdyenEncryption
import Foundation

final class AdyenCSE {
    func encryptCard(
        unencryptedCardDTO: UnencryptedCardDTO,
        publicKey: String
    ) -> Result<EncryptedCardDTO, Error> {
        do {
            let encryptedCard = try CardEncryptor.encrypt(
                card: Card(
                    number: unencryptedCardDTO.cardNumber,
                    securityCode: unencryptedCardDTO.cvc,
                    expiryMonth: unencryptedCardDTO.expiryMonth,
                    expiryYear: unencryptedCardDTO.expiryYear
                ),
                with: publicKey
            )
            return .success(
                EncryptedCardDTO(
                    encryptedCardNumber: encryptedCard.number,
                    encryptedExpiryMonth: encryptedCard.expiryMonth,
                    encryptedExpiryYear: encryptedCard.expiryYear,
                    encryptedSecurityCode: encryptedCard.securityCode
                )
            )
        } catch {
            return .failure(error)
        }
    }

    func encryptBin(bin: String, publicKey: String) -> Result<String, Error> {
        do {
            return try .success(CardEncryptor.encrypt(bin: bin, with: publicKey))
        } catch {
            return .failure(error)
        }
    }
}
